### ISSUE TITLE
Post Processing Bottleneck on Jetson AGX Orin

### ISSUE THREAD

**Issue by TristanPuma1** [ultralytics/ultralytics#17542]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hello, I'm also getting unexpectedly high post-processing times, inferencing on a 64GB Jetson AGX Orin. After exporting the model to an engine format, tensor appears to be successfully inferencing on the GPU but post processing tasks are not. 

Here is code executing the export in addition to its output. Below that I also attempted to load the engine directly into tensor so that I can have more manual control of memory allocation, but I'm getting a serialization error. If anyone has any idea of what might be causing this post-processing bottleneck or ways that I could address it please let me know! Are there ultralytics scripts that would allow me to breakup the preprocess, inference and postprocess steps? 

from ultralytics import YOLO

model = YOLO("yolov8s.pt")
model.export(format="engine", device=0)
trt_model = YOLO("yolov8s.engine")

# Inference 
results = trt_model("bus.jpg")

_______________________________________________________________________________________________________________________________________________________

(mask_env) tristan@ubuntu:~/Downloads/Models_Imported$ python export_to_engine.py
Ultralytics 8.3.29 🚀 Python-3.10.12 torch-2.3.0 CUDA:0 (Orin, 62841MiB)
YOLOv8s summary (fused): 168 layers, 11,156,544 parameters, 0 gradients, 28.6 GFLOPs

PyTorch: starting from 'yolov8s.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) (1, 84, 8400) (21.5 MB)

ONNX: starting export with onnx 1.17.0 opset 17...
ONNX: slimming with onnxslim 0.1.39...
ONNX: export success ✅ 1.9s, saved as 'yolov8s.onnx' (42.8 MB)

TensorRT: starting export with TensorRT 8.6.2...
[11/14/2024-12:51:57] [TRT] [I] [MemUsageChange] Init CUDA: CPU +2, GPU +0, now: CPU 731, GPU 5433 (MiB)
[11/14/2024-12:52:01] [TRT] [I] [MemUsageChange] Init builder kernel library: CPU +1154, GPU +1358, now: CPU 1921, GPU 6778 (MiB)
[11/14/2024-12:52:01] [TRT] [I] ----------------------------------------------------------------
[11/14/2024-12:52:01] [TRT] [I] Input filename:   yolov8s.onnx
[11/14/2024-12:52:01] [TRT] [I] ONNX IR version:  0.0.8
[11/14/2024-12:52:01] [TRT] [I] Opset version:    17
[11/14/2024-12:52:01] [TRT] [I] Producer name:    pytorch
[11/14/2024-12:52:01] [TRT] [I] Producer version: 2.3.0
[11/14/2024-12:52:01] [TRT] [I] Domain:           
[11/14/2024-12:52:01] [TRT] [I] Model version:    0
[11/14/2024-12:52:01] [TRT] [I] Doc string:       
[11/14/2024-12:52:01] [TRT] [I] ----------------------------------------------------------------
[11/14/2024-12:52:01] [TRT] [W] onnx2trt_utils.cpp:372: Your ONNX model has been generated with INT64 weights, while TensorRT does not natively support INT64. Attempting to cast down to INT32.
TensorRT: input "images" with shape(1, 3, 640, 640) DataType.FLOAT
TensorRT: output "output0" with shape(1, 84, 8400) DataType.FLOAT
TensorRT: building FP32 engine as yolov8s.engine
[11/14/2024-12:52:01] [TRT] [I] Graph optimization time: 0.03221 seconds.
[11/14/2024-12:52:01] [TRT] [I] Local timing cache in use. Profiling results in this builder pass will not be stored.
[11/14/2024-12:56:05] [TRT] [I] Detected 1 inputs and 3 output network tensors.
[11/14/2024-12:56:05] [TRT] [I] Total Host Persistent Memory: 369984
[11/14/2024-12:56:05] [TRT] [I] Total Device Persistent Memory: 0
[11/14/2024-12:56:05] [TRT] [I] Total Scratch Memory: 0
[11/14/2024-12:56:05] [TRT] [I] [MemUsageStats] Peak memory usage of TRT CPU/GPU memory allocators: CPU 10 MiB, GPU 260 MiB
[11/14/2024-12:56:05] [TRT] [I] [BlockAssignment] Started assigning block shifts. This will take 168 steps to complete.
[11/14/2024-12:56:05] [TRT] [I] [BlockAssignment] Algorithm ShiftNTopDown took 10.333ms to assign 8 blocks to 168 nodes requiring 38912512 bytes.
[11/14/2024-12:56:05] [TRT] [I] Total Activation Memory: 38912000
[11/14/2024-12:56:05] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in building engine: CPU +6, GPU +64, now: CPU 6, GPU 64 (MiB)
TensorRT: export success ✅ 251.4s, saved as 'yolov8s.engine' (47.4 MB)

Export complete (252.6s)
Results saved to /home/tristan/Downloads/Models_Imported
Predict:         yolo predict task=detect model=yolov8s.engine imgsz=640  
Validate:        yolo val task=detect model=yolov8s.engine imgsz=640 data=coco.yaml  
Visualize:       https://netron.app
WARNING ⚠️ Unable to automatically guess model task, assuming 'task=detect'. Explicitly define task for your model, i.e. 'task=detect', 'segment', 'classify','pose' or 'obb'.
Loading yolov8s.engine for TensorRT inference...
[11/14/2024-12:56:06] [TRT] [I] Loaded engine size: 47 MiB
[11/14/2024-12:56:06] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in engine deserialization: CPU +0, GPU +42, now: CPU 0, GPU 42 (MiB)
[11/14/2024-12:56:06] [TRT] [I] [MemUsageChange] TensorRT-managed allocation in IExecutionContext creation: CPU +0, GPU +37, now: CPU 0, GPU 79 (MiB)

image 1/1 /home/tristan/Downloads/Models_Imported/bus.jpg: 640x640 4 persons, 1 bus, 21.7ms
Speed: 6.7ms preprocess, 21.7ms inference, 507.1ms postprocess per image at shape (1, 3, 640, 640)

_______________________________________________________________________________________________________________________________________________________

import tensorrt as trt
import pycuda.driver as cuda
import pycuda.autoinit
import numpy as np
import cv2
import time
import os

TRT_LOGGER = trt.Logger(trt.Logger.WARNING)         # init trt logger

# Load engine
def load_engine(engine_path):
    # Check if path exists
    if not os.path.exists(engine_path):
        raise FileNotFoundError(f"Engine file not found at: {engine_path}")
    else:
        print("file path valid")

    with open(engine_path, "rb") as f, trt.Runtime(TRT_LOGGER) as runtime:
        return runtime.deserialize_cuda_engine(f.read())
    

engine_path = "models/yolov8s_fp32/yolov8s.engine"
engine = load_engine(engine_path)

[11/14/2024-13:37:42] [TRT] [E] 1: [runtime.cpp::parsePlan::314] Error Code 1: Serialization (Serialization assertion plan->header.magicTag == rt::kPLAN_MAGIC_TAG failed.)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @TristanPuma1, thank you for reaching out to the Ultralytics community 🚀! We're thrilled to help you with your inquiry regarding the post-processing bottleneck on the Jetson AGX Orin.

We recommend visiting the [Docs](https://docs.ultralytics.com) for detailed guidance on various topics, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage, which might offer insights into common issues like the one you're experiencing.

Since it seems like there may be an underlying issue with the post-processing, could you please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) along with any relevant error logs? This would greatly assist our team in diagnosing the problem effectively.

If this is a custom training ❓ question, ensure to share detailed information, including dataset samples and training logs, while checking you're aligned with our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Meanwhile, our vibrant Ultralytics community is ready to engage! For real-time chat, join us on [Discord](https://ultralytics.com/discord) 🎧. For comprehensive discussions, explore [Discourse](https://community.ultralytics.com). Or share your journey and insights on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you're using the latest version of the `ultralytics` package with all required dependencies. We recommend a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure all patches and updates are in effect:

```bash
pip install -U ultralytics
```

## Environments

YOLO can operate efficiently in any of the following up-to-date environments with full dependencies installed, including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks with free GPU:** <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing. These tests ensure all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) are functioning correctly across various operating systems.

Thank you for your patience as you await further assistance from an Ultralytics engineer! 😊

**Comment by glenn-jocher** [comment]
@TristanPuma1 thank you for providing such detailed information about your post-processing bottleneck issue on the Jetson AGX Orin. Based on the timing breakdown you shared (6.7ms preprocess, 21.7ms inference, 507.1ms postprocess), there's indeed a significant bottleneck in post-processing.

For optimizing post-processing on Jetson platforms, I recommend using TensorRT with DLA (Deep Learning Accelerator) and enabling FP16 precision. Here's a modified export command that should help:

```python
model.export(format="engine", device="dla:0", half=True, workspace=4)
```

For manual control over the processing pipeline, you can use YOLO's predictor interface:

```python
from ultralytics.engine.predictor import BasePredictor
predictor = BasePredictor(overrides={'model': 'yolov8s.engine'})
```

For additional Jetson-specific optimizations and benchmarks, please refer to our Jetson guide at https://docs.ultralytics.com/guides/nvidia-jetson/.

Regarding the TensorRT serialization error, this typically occurs due to version mismatches. Ensure you're using compatible versions of TensorRT between export and inference environments.

**Comment by Y-T-G** [comment]
> Are there ultralytics scripts that would allow me to breakup the preprocess, inference and postprocess steps?

You can use this

https://github.com/ultralytics/ultralytics/issues/14855#issuecomment-2262134660
